File: app/src/main/java/no/nav/tilbakemeldingsmottak/rest/bestillingavsamtale/BestillingAvSamtaleRestController.java
package no.nav.tilbakemeldingsmottak.rest.bestillingavsamtale;

import static no.nav.tilbakemeldingsmottak.metrics.MetricLabels.DOK_REQUEST;
import static no.nav.tilbakemeldingsmottak.metrics.MetricLabels.PROCESS_CODE;

import lombok.extern.slf4j.Slf4j;
import no.nav.security.token.support.core.api.ProtectedWithClaims;
import no.nav.tilbakemeldingsmottak.consumer.email.SendEmailException;
import no.nav.tilbakemeldingsmottak.metrics.Metrics;
import no.nav.tilbakemeldingsmottak.model.BestillSamtaleResponse;
import no.nav.tilbakemeldingsmottak.model.BestillSamtaleRequest;
import no.nav.tilbakemeldingsmottak.rest.bestillingavsamtale.service.BestillingAvSamtaleService;
import no.nav.tilbakemeldingsmottak.rest.bestillingavsamtale.validation.BestillSamtaleValidator;
import no.nav.tilbakemeldingsmottak.api.BestillingAvSamtaleRestControllerApi;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import javax.inject.Inject;
import javax.transaction.Transactional;

@Slf4j
@ProtectedWithClaims(issuer = "azuread")
@RestController
@CrossOrigin(maxAge = 3600)
public class BestillingAvSamtaleRestController implements BestillingAvSamtaleRestControllerApi {

    private final BestillingAvSamtaleService bestillingAvSamtaleService;
    private final BestillSamtaleValidator bestillSamtaleValidator;


    @Inject
    public BestillingAvSamtaleRestController(final BestillingAvSamtaleService bestillingAvSamtaleService) {
        this.bestillingAvSamtaleService = bestillingAvSamtaleService;
        this.bestillSamtaleValidator = new BestillSamtaleValidator();
    }

    @Transactional
    @Override
    @Metrics(value = DOK_REQUEST, extraTags = {PROCESS_CODE, "sendRos"}, percentiles = {0.5, 0.95}, histogram = true)
    public ResponseEntity<BestillSamtaleResponse> bestillingAvSamtale(@RequestBody BestillSamtaleRequest request) throws SendEmailException {
            bestillSamtaleValidator.validateRequest(request);
            bestillingAvSamtaleService.bestillSamtale(request);
            return ResponseEntity
                    .status(HttpStatus.OK)
                    .body(BestillSamtaleResponse.builder().message("Samtale bestilt").build());
    }
}

File: app/src/main/kotlin/no/nav/tilbakemeldingsmottak/rest/bestillingavsamtale/BestillingAvSamtaleRestController.kt
package no.nav.tilbakemeldingsmottak.rest.bestillingavsamtale

import no.nav.security.token.support.core.api.ProtectedWithClaims
import no.nav.tilbakemeldingsmottak.api.BestillingAvSamtaleRestControllerApi
import no.nav.tilbakemeldingsmottak.config.Constants
import no.nav.tilbakemeldingsmottak.metrics.MetricLabels
import no.nav.tilbakemeldingsmottak.metrics.Metrics
import no.nav.tilbakemeldingsmottak.metrics.MetricsUtils
import no.nav.tilbakemeldingsmottak.model.BestillSamtaleRequest
import no.nav.tilbakemeldingsmottak.model.BestillSamtaleResponse
import no.nav.tilbakemeldingsmottak.rest.bestillingavsamtale.service.BestillingAvSamtaleService
import no.nav.tilbakemeldingsmottak.rest.bestillingavsamtale.validation.BestillSamtaleValidator
import no.nav.tilbakemeldingsmottak.util.OidcUtils
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController

@ProtectedWithClaims(issuer = "azuread")
@RestController
class BestillingAvSamtaleRestController(
    private val bestillingAvSamtaleService: BestillingAvSamtaleService,
    private val metricsUtils: MetricsUtils,
    private val oidcUtils: OidcUtils
) : BestillingAvSamtaleRestControllerApi {

    private val bestillSamtaleValidator: BestillSamtaleValidator = BestillSamtaleValidator()

    @Transactional
    @Metrics(
        value = MetricLabels.DOK_REQUEST,
        extraTags = [MetricLabels.PROCESS_CODE, "bestill-samtale"],
        percentiles = [0.5, 0.95],
        histogram = true
    )
    override fun bestillingAvSamtale(@RequestBody bestillSamtaleRequest: BestillSamtaleRequest): ResponseEntity<BestillSamtaleResponse> {
        if (oidcUtils.getPidForIssuer(Constants.TOKENX_ISSUER) == null) {
            metricsUtils.incrementNotLoggedInRequestCounter(this.javaClass.name, "bestillingAvSamtale")
        }
        bestillSamtaleValidator.validateRequest(bestillSamtaleRequest)
        bestillingAvSamtaleService.bestillSamtale(bestillSamtaleRequest)
        return ResponseEntity
            .status(HttpStatus.OK)
            .body(BestillSamtaleResponse(message = "Samtale bestilt"))
    }
}
